pkg/serving/v1/gitops.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2020 The Knative Authors
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+
+	"knative.dev/client/pkg/wait"
+
+	"k8s.io/apimachinery/pkg/watch"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+const (
+	operationNotSuportedError = "this operation is not supported in gitops mode"
+	ksvcKind                  = "ksvc"
+)
+
+// knServingGitOpsClient - kn service client
+// to work on a local repo instead of a remote cluster
+type knServingGitOpsClient struct {
+	dir        string
+	namespace  string
+	fileClient fileOpsClient
+}
+
+type fileOpsClient interface {
+	writeToFile(string) (io.Writer, error)
+	readFromFile(string) (io.Reader, error)
+	removeFile(string) error
+	listFiles(string) ([]servingv1.Service, error)
+}
+
+type fileClient struct{}
+
+// NewKnServingGitOpsClient returns an instance of the
+// kn service gitops client
+func NewKnServingGitOpsClient(namespace, dir string) *knServingGitOpsClient {
+	// create dir , if not present
+	namespaceDir := filepath.Join(dir, namespace, ksvcKind)
+	if _, err := os.Stat(namespaceDir); os.IsNotExist(err) {
+		os.MkdirAll(namespaceDir, 0777)
+	}
+	return &knServingGitOpsClient{
+		dir:        dir,
+		namespace:  namespace,
+		fileClient: &fileClient{},
```

Comment:
the methods in the client all have direct equivalents to system libraries, so maybe even the functions are not needed (and you could use the system calls directly ?)

Reply:
have this so that i could create fake file client for testing. For unit tests i dint want to write files to directories.

Reply:
tbh, I'm not a big fan of adapting the original code so that you can better test it (same for making private methods public or package visible in Java, just so that the test can access so internal variables). What's about adapting the test to let them write into a temporary directory, and then validate on the content of the written files? (and then delete the directory after the test)This would be also a little bit more realistic as it includes the same filesystem interaction as for the real business case. Also, I don't think its much slower.

It's maybe not a "pure" unit test, but it would also help us easily to detect issues on operating systems (e.g. windows) easier as long as we don't have full e2e test coverage for all the platforms that we support (e.g. linux, windows, macos, arm64, ....)

I really would avoid that extra abstraction introduced only for the sake of testing. (tl;dr - this is a 'leaky' abstraction is introduced to overcome technical constraints but shines through to the implementation of the business logic)

Reply:
Sure Roland, that was my initial plan (create file and delete it) but i was not sure on where to draw the line between e2e and unit testing. You are right, if you see my initial commit it dint have this abstraction layer, it was introduced for unit testing.

Reply:
+1, thanks for taking care. Normally I would totally agree with you and filesystem interactions are not at the heart of unit tests. But one should not be dogmatic here, and it it takes a lot of extra efforts that leak into the code to test, then I think one should be just pragmatic ;-) With accessing the filesystem this might be often the case. Of course it would be much different when talking e.g. about databases or so.

Reply:
have made the changes. Please review